### PR TITLE
cmake: running unit tests

### PR DESCRIPTION
**Pull request by furiel**
Execute in parallel: make test ARGS=-j4

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
-
+add_subdirectory(unit)
 add_subdirectory(loggen)
-
+# add_subdirectory(functional)
```
Comment (Kokan) [inline tests/CMakeLists.txt]: Should be removed.

File: lib/filter/tests/test_filters_in_list.c [hunk 2]
```diff
@@ -157,7 +157,11 @@ run_testcases(const char *top_srcdir)
 int
 main(int argc G_GNUC_UNUSED, char *argv[] G_GNUC_UNUSED)
 {
+#ifdef TOP_SRCDIR
```
Comment (Kokan) [inline lib/filter/tests/test_filters_in_list.c]: I like your approach, but I think it would be better, if the `autotools` based build would use the same approach.
Could you please update the `autotools` compile also ?

File: lib/logproto/tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+set(TEST_LOGPROTO_SOURCES
+  test_logproto.c
+  test_logproto.h
+  test-server-options.c
+  test-record-server.c
+  test-text-server.c
+  test-dgram-server.c
+  test-framed-server.c
+  test-indented-multiline-server.c
+  test-regexp-multiline-server.c)
```
[the remark this answers is not in the document]
Comment (furiel) [reply lib/logproto/tests/CMakeLists.txt]: fixed

File: modules/dbparser/tests/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+add_unit_test(TARGET test_timer_wheel INCLUDES ${PATTERNDB_INCLUDE_DIR} DEPENDS patterndb)
+add_unit_test(TARGET test_patternize INCLUDES ${PATTERNDB_INCLUDE_DIR} DEPENDS patterndb syslogformat)
+add_unit_test(LIBTEST TARGET test_patterndb INCLUDES ${PATTERNDB_INCLUDE_DIR}
+  DEPENDS patterndb basicfuncs syslogformat)
+add_unit_test(TARGET test_radix INCLUDES ${PATTERNDB_INCLUDE_DIR} DEPENDS patterndb)
+target_compile_options(test_radix PRIVATE "-Wno-pointer-sign")
```
[the remark this answers is not in the document]
Comment (furiel) [reply modules/dbparser/tests/CMakeLists.txt]: fixed

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: modules/json/CMakeLists.txt [hunk 5]
```diff
@@ -12,6 +12,8 @@ set(JSON_SOURCES
     ${CMAKE_CURRENT_BINARY_DIR}/json-parser-grammar.c
 )
 
+set(SJON_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (MrAnno) [inline modules/json/CMakeLists.txt]: typo

File: modules/json/tests/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,19 @@
+add_unit_test(LIBTEST TARGET test_format_json
+  DEPENDS syslogformat json-plugin ${JSONC_LIBRARY})
+if (${JSONC_INTERNAL})
+  add_dependencies(test_format_json JSONC)
+endif()
+
+add_unit_test(LIBTEST TARGET test_json_parser
+  INCLUDES "${SJON_INCLUDE_DIR}"
```
Comment (MrAnno) [inline modules/json/tests/CMakeLists.txt]: typo

File: modules/json/tests/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  DEPENDS json-plugin ${JSONC_LIBRARY})
+if (${JSONC_INTERNAL})
+  add_dependencies(test_json_parser JSONC)
+endif()
+
+add_unit_test(LIBTEST TARGET test_dot_notation
+  INCLUDES "${SJON_INCLUDE_DIR}" "${JSONC_INCLUDE_DIR}"
```
Comment (MrAnno) [inline modules/json/tests/CMakeLists.txt]: typo

File: cmake/Modules/ExternalJSONC.cmake [hunk 8]
```diff
@@ -48,13 +48,10 @@ if (EXISTS ${PROJECT_SOURCE_DIR}/lib/jsonc/json.h)
      )
 
     set(${LIB_NAME}_INTERNAL TRUE)
-    set(${LIB_NAME}_INTERNAL_INCLUDE_DIR "${PROJECT_BINARY_DIR}/include/json-c/" INTERNAL)
-    set(${LIB_NAME}_INTERNAL_LIBRARY "${PROJECT_BINARY_DIR}/lib/libjson-c.so" INTERNAL)
+    set(${LIB_NAME}_INTERNAL_INCLUDE_DIR "${PROJECT_BINARY_DIR}/include/json-c/")
```
Comment (MrAnno) [inline cmake/Modules/ExternalJSONC.cmake]: Is this related to the unit tests?
If not, can you extract the changes of this file to a separate commit?

File: cmake/add_tests.cmake [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+
+
+include(CMakeParseArguments)
+
+function (add_unit_test)
```
Comment (MrAnno) [inline cmake/add_tests.cmake]: Nice :+1:

File: cmake/Modules/Findsystemd.cmake [hunk 10]
```diff
@@ -30,5 +30,4 @@ set(Libsystemd_PROCESS_INCLUDES Libsystemd_INCLUDE_DIR)
 set(Libsystemd_PROCESS_LIBS Libsystemd_LIBRARY)
 
 # silence warnings
-set(Libsystemd_FIND_QUIETLY 1)
```
Comment (MrAnno) [inline cmake/Modules/Findsystemd.cmake]: Can you extract this change to a separate commit?

File: lib/filter/tests/test_filters_in_list.c [hunk 11]
```diff
@@ -157,7 +157,8 @@ run_testcases(const char *top_srcdir)
 int
 main(int argc G_GNUC_UNUSED, char *argv[] G_GNUC_UNUSED)
 {
-  char *top_srcdir = getenv("top_srcdir");
+
+  char *top_srcdir = TOP_SRCDIR;
```
Comment (MrAnno) [inline lib/filter/tests/test_filters_in_list.c]: I would split this commit (along with the Makefile change).
In my opinion, this is a bugfix-like patch and has nothing to do with CMake.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: cmake/Modules/ExternalJSONC.cmake (hunk 8 above)
Comment (furiel) [reply cmake/Modules/ExternalJSONC.cmake]: Fixed

File: cmake/Modules/Findsystemd.cmake (hunk 10 above)
Comment (furiel) [reply cmake/Modules/Findsystemd.cmake]: Fixed
